### DISCUSSION TITLE
Valhalla on GPUs ?

### DISCUSSION THREAD

**Discussion by answerquest** [valhalla/valhalla#4984] [Q&A]
Apologies in advance for being very low-information on this.
There's a lot of recent advancements in GPU based processing for AI applications.
My very basic understanding is that a problem is broken down into several independent parts and they're given over to the GPUs to compute in parallel.

I'm not sure if what valhalla does can fit into this, but for matrix api we can split the problem into smaller parts and have them done in parallel.

Is there a way to run Valhalla on a system with GPUs and take advantage of those resources?

**Comment by kevinkreiser** [comment]
that sounds like a tall order to me. back in the day i wrote cuda gpgpu programs and glsl/hlsl shader stuff for graphics applications and if that is how gpu programming is still done this day in age then id have a very hard time seeing valhalla ever run on a gpu without essentially a rewrite.

not saying its a bad idea but i think if we wanted to target gpus 1 of two approaches would have to be taken:

1. some magic compiler that can target gpu architecture with non-gpu specific code
2. write a new router thats specifically designed to target gpus from its inception

interesting question but im afraid i dont have anything promising to say about it, maybe other experts in modern day gpgpu will have somethign to say. i would also say that fundamentally im not sure gpus really make sense for conventional routing algorithms in a practical sense. i do think you can write parallelized a* that targets gpus but in practice, that is a usable algorithm for consumer grade routing, needs to evaluate the cost of various edges and that computation requires access to a ton of data. sure you could cram a bunch of this into textures but there is variable size data and if you add in a time of day component you're really screwed (cant do that in parallel). so in short, not impossible but sounds highly impractical to me!
